shakenfist/scheduler.py

```diff
@@ -239,73 +240,89 @@ def place_instance(self, instance, network, candidates=None):
-            instance.add_event('schedule',
-                               'Have enough idle CPU', None, str(candidates))
+            inst.add_event('schedule',
+                           'Have enough idle CPU', None, str(candidates))
             if not candidates:
                 raise exceptions.LowResourceException(
                     'No nodes with enough idle CPU')
 
             # Do we have enough idle RAM?
             for n in copy.copy(candidates):
-                if not self._has_sufficient_ram(log_ctx, instance.memory, n):
+                if not self._has_sufficient_ram(log_ctx, inst.memory, n):
                     candidates.remove(n)
             log_ctx.with_field('candidates', candidates).info(
                 'Scheduling: Have enough idle RAM')
-            instance.add_event('schedule',
-                               'Have enough idle RAM', None, str(candidates))
+            inst.add_event('schedule',
+                           'Have enough idle RAM', None, str(candidates))
             if not candidates:
                 raise exceptions.LowResourceException(
                     'No nodes with enough idle RAM')
 
             # Do we have enough idle disk?
             for n in copy.copy(candidates):
-                if not self._has_sufficient_disk(log_ctx, instance, n):
+                if not self._has_sufficient_disk(log_ctx, inst, n):
                     candidates.remove(n)
             log_ctx.with_field('candidates', candidates).info(
                 'Scheduling: Have enough idle disk')
-            instance.add_event('schedule',
-                               'Have enough idle disk', None, str(candidates))
+            inst.add_event('schedule',
+                           'Have enough idle disk', None, str(candidates))
             if not candidates:
                 raise exceptions.LowResourceException(
                     'No nodes with enough disk space')
 
+            # Calc pseudo CPU load from inst affinity
+            pseudo_load = defaultdict(int)
+            cpu_affinity = inst.affinity.get('cpu')
+            if cpu_affinity:
+                for i in instance.Instances([instance.healthy_states_filter]):
+                    if not i.tags:
+                        continue
+                    for tag, val in cpu_affinity.items():
+                        if tag in i.tags:
+                            # Use get() to allow for unplaced healthy instances
+                            pseudo_load[i.placement.get('node')] += val
+
+            inst.add_event('schedule', 'Pseudo load due inst affinity',
+                           None, str(dict(pseudo_load)))
+
             # Order candidates by current CPU load
-            by_load = {}
+            by_load = defaultdict(list)
             for n in copy.copy(candidates):
                 load = math.floor(self.metrics[n].get('cpu_load_1', 0))
-                by_load.setdefault(load, [])
+                load -= pseudo_load.get(n, 0)
```

Comment:
Why do we substract the pseudo load here? I would of thought we added it?

Reply:
The value supplied is affinity. The node with the lowest load is chosen, therefore if an instance has affinity with another instance it has to reduce the CPU load on a node in order to be placed on that node.

It follows that an instance expresses anti-affinity with a negative number. `--cpuaffinity nastyvm=-99`

(I've written very little doc since I'm just floating the idea and seeing if passes CI.)

Reply:
I don't know how I feel about using CPU load here. Doesn't that mean an affinity might result in you landing on a very overloaded node? I feel like maybe this should be a separate filter?

Reply:
Remember that this filter is after the `_has_sufficient_cpu` filter, so complete overcommit is not possible.

If this is a separate filter, then the user loses the ability to say "I want this instance away from that other instance, only if the non-optimal CPU load imbalance is less than this value".

I would suggest, implement this `CPU affinity` (and `disk space affinity`) and also implement a pure affinity that is actually a binary decision to share a node with a particular tag.

This is only "affinity" not outright placement "control". The affinity value should be low to express "I desire to be close to another instance". To express "I demand" then it can be a value higher than the max possible CPU load.

The user might know that the load will reduce after instance startup, so is happy with initial higher load. This is about letting the user utilise their knowledge and communicate what is best, without the user demanding a placement that would be detrimental.

Reply:
Does that imply that for me to specify "I demand", I need to know what the max CPU load for a given node is, or do I just use a really big number like 100?

Reply:
Correct. Use a really big number. Turns the system in a control system rather than affinity - allowing for overcommit limits.

In reality, forcing an instance on or off a node when the CPU load advantage is greater than an entire physical CPU seems counter-productive.